# A/B Testing Simulation

## Introduction

A/B testing is a common technique used to test whether a new design or feature improves the performance of a website compared to the previous one. Here, we will be performing A/B testing to determine whether a new website design increases the conversion rate compared to the old design. 

We randomly divide 2000 website visitors into two groups: 1000 visitors see the old design (group A) and 1000 visitors see the new design (group B). We assume that the true conversion rates are 10% for group A and 12% for group B.

In the context of A/B testing, a [conversion rate](https://mailchimp.com/marketing-glossary/conversion-rates/) typically measures the percentage of users who completed a desired action, such as making a purchase or signing up for a service, out of the total number of users who were presented with the opportunity to take that action.

For example, if 1,000 people visited a website and 50 of them made a purchase, the conversion rate would be 5% (50/1000). Similarly, if 500 people were shown an advertisement and 20 of them clicked on the ad, the conversion rate would be 4% (20/500).

***
## Designing the experiment

### Formulating a hypothesis
We'll choose a <a href="https://www.ncl.ac.uk/webtemplate/ask-assets/external/maths-resources/statistics/hypothesis-testing/one-tailed-and-two-tailed-tests.html#:~:text=The%20main%20difference%20between%20one,using%20a%20two%2Dtailed%20test.">**one-tailed test**</a>:

$$H_0: p = p_0$$
$$H_a: p > p_0$$

In words; <br>

Null Hypothesis: There is no significant difference in the conversion rates between group A and group B.

Alternative Hypothesis: The conversion rate of group B is significantly higher than the conversion rate of group A.

where $p$ and $p_0$ stand for the conversion rate of the new and old design, respectively. 
We'll set a **confidence level of 95%**:

$$\alpha = 0.05$$

The $\alpha$ value is the probability of rejecting the null hypothesis when it is actually true. A common significance level is 0.05, which means that there is a 5% chance of incorrectly rejecting the null hypothesis. Since our $\alpha=0.05$ (indicating 5% probability), our confidence (1 - $\alpha$) is 95%.

### The variables

For our test, we need two groups: Group A and Group B. Group A is the control group, and they will be shown the old design. Group B is the treatment group, and they will be shown the new design. We are interested in capturing the conversion rate, which we will use as our dependent variable. We can measure the conversion rate by assigning a binary variable to each user session. The binary variable will be 1 if the user bought the product during that user session and 0 if they did not buy the product.
* `0` - The user did not buy the product during this user session
* `1` - The user bought the product during this user session

In [1]:
# Import required libraries
import numpy as np
from scipy.stats import ttest_ind

In [2]:
# Set random seed for reproducibility
np.random.seed(40)

In [3]:
# Define sample size and true conversion rates
n = 1000
rate_A = 0.10
rate_B = 0.12

We can simulate the results of the A/B test by randomly generating binary outcomes for each visitor: 0 for no conversion and 1 for conversion. If the conversion rate is 10%, we can use 0.1 to generate a random binary outcome.We repeat this process for each visitor in both groups to obtain two arrays of binary outcomes, representing the conversion results for group A and group B.

In [4]:
# Generate binary outcomes for group A
outcomes_A = np.random.binomial(n=1, p=rate_A, size=n)

# Generate binary outcomes for group B
outcomes_B = np.random.binomial(n=1, p=rate_B, size=n)

Next, we can calculate the conversion rates and perform hypothesis testing to determine whether the difference between the two rates is statistically significant.

In [5]:
conv_rate_A = outcomes_A.mean()
conv_rate_B = outcomes_B.mean()

print(f"Conversion rate for group A: {conv_rate_A:.2%}")
print(f"Conversion rate for group B: {conv_rate_B:.2%}")

Conversion rate for group A: 10.00%
Conversion rate for group B: 13.20%


In [6]:
# Perform hypothesis testing
t_stat, p_value = ttest_ind(outcomes_A, outcomes_B)

print(f"T-statistic: {t_stat:.2f}")
print(f"P-value: {p_value:.3f}")

T-statistic: -2.24
P-value: 0.025


We have generated binary outcomes for group A and group B by randomly generating binary outcomes for each visitor. We have repeated this process for each visitor in both groups to obtain two arrays of binary outcomes representing the conversion results for group A and group B. We have calculated the conversion rates and performed hypothesis testing to determine whether the difference between the two rates is statistically significant.

In [7]:
# Check significance level
alpha = 0.05
if p_value < alpha:
    print("Reject null hypothesis")
else:
    print("Fail to reject null hypothesis")


Reject null hypothesis


In [8]:
# Calculate confidence interval
se = np.sqrt(conv_rate_A * (1 - conv_rate_A) / n + conv_rate_B * (1 - conv_rate_B) / n)
margin_error = 1.96 * se
ci_low = (conv_rate_B - conv_rate_A) - margin_error
ci_high = (conv_rate_B - conv_rate_A) + margin_error
print(f"95% Confidence Interval: [{ci_low:.2%}, {ci_high:.2%}]")

95% Confidence Interval: [0.40%, 6.00%]


## Results 
We found that the conversion rate for group A was 10%, while the conversion rate for group B was 13.2%. This means that the conversion rate of group B is higher than the conversion rate of group A. We have performed hypothesis testing to confirm whether the difference between the conversion rates of group A and group B is statistically significant. The p-value we obtained from the test was 0.025, which is less than the significance level of 0.05. Therefore, we can reject the null hypothesis and conclude that the conversion rate of group B is significantly higher than the conversion rate of group A.

## Conclusion
In conclusion, we have successfully performed A/B testing to determine whether a new website design increases the conversion rate compared to the old design. We have found that the new design is more effective in increasing the conversion rate than the old design. This suggests that implementing the new design on the website will likely result in a higher conversion rate, and therefore, more sales.